lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Delegation.hs

```diff
@@ -0,0 +1,464 @@
+-- States caused by                             └────┐Pointer                         ▼       │       ┌─┘     └─┐       ▼       │        ┌─┘     └─┐
+-- old wallet                                        └deposit                  ┌────────────────────┐─┘         └┬─────────────────────┐─┘         └─
+-- de-registering                                     returned─┐               │                    │            │                     │
+-- stake-key 0                                                 └────┐          │                    │ ─────────▶ │                     │ ─────────▶
+-- of multi-stake                                                   └────┐     │                    │            │                     │
+-- key wallet                                                            └──── │        [1]         │            │        [1,2]        │
+--                                                                             │1 ada held by key 2 │            │ 1 ada held by key 3 │
+--                                                                             │                    │            │                     │
+--                                                                             │                    │◀────────── │                     │◀──────────
+--                                                                             │                    │            │                     │
+--                                                                             │                    │            │                     │
+--                                                                             └────────────────────┘            └─────────────────────┘
+-- @
+data DelegationState k = DelegationState
+    { -- | The account public key from which the stake keys should be derived.
+      rewardAccountKey :: k 'AccountK XPub
+    , state :: State
+    } deriving (Generic)
+
+
+
+-- | The internal state, without the account key.
+--
+-- TODO: Perhaps we should model this as
+-- @S = S1 * S2, where S1 = Bool, S2 = ix * UTxOPointer@ - having two
+-- "concurrent" states tracking stake keys, where the first one is identical to
+-- legacy single-stake key wallets.
+--
+-- Maybe that would help simplify `applyTx` and `setPortfolioOf`...
+data State
+    -- | No active stake keys. The initial state of a new wallet.
+    = Zero
+
+    -- | The first stake-key (index 0) is registered and either delegating or
+    -- about to be delegating.
+    | One
+
+    -- | There is more than one active stake keys. Can only be reached using
+    -- wallets with support for multiple stake keys.
+    | More
+        !(Index 'Soft 'AddressK)
+          -- nextKeyIx - the ix of the next unused key
+        PointerUTxO
+          -- ^ pointer utxo that need to be spent when changing state.
+        !Bool
+          -- ^ Is key 0 still registered? For compatibility with
```

Comment:
A little newtype or sum-type would be better :pray:

Reply:
I went with

```haskell
-- | Is key 0 still registered? For compatibility with
-- single-stake-key wallets, we need to track this.
--
-- >>> activeKeys (More (toEnum 3) p ValidKey0)
-- [0, 1, 2]
--
-- >>> activeKeys (More (toEnum 3) p MissingKey0)
-- [1, 2]
--
-- (pseudocode; requires a bit more boilerplate to compile)
--
-- See the implementaiton of @applyTx@ for how it is used.
data Key0Status = ValidKey0 | MissingKey0
    deriving (Eq, Show, Generic)
```
ce5a3718e9607885b84c370badd75a0d93d4a6bc